R/Rules-methods.R

```diff
@@ -284,81 +284,58 @@ setMethod(
+#'
+#' @aliases nextBest-NextBestThreePlusThree
+#'
+#' @export
+#' @example examples/Rules-method-nextBest-NextBestThreePlusThree.R
+#'
+setMethod(
+  f = "nextBest",
+  signature = signature(
+    nextBest = "NextBestThreePlusThree",
+    doselimit = "missing",
+    samples = "missing",
+    model = "missing",
+    data = "Data"
+  ),
+  definition = function(nextBest, doselimit, samples, model, data, ...) {
 
-              ## number of patients and number of DLTs per dose level group
-              nPatients <- sapply(dltSplit, length)
-              nDLTs <- sapply(dltSplit, sum)
+    # The last dose level tested (not necessarily the maximum one).
+    last_level <- tail(data@xLevel, 1L)
 
-              ## what was the last dose level tested?
-              lastLevel <- tail(data@xLevel, 1)
+    # Get number of patients per grid's dose and DLT rate at the last level.
+    nPatients <- table(factor(data@x, levels = data@doseGrid))
+    nDLTs_last_level <- sum(data@y[data@xLevel == last_level])
+    DLT_rate_last_level <- nDLTs_last_level / nPatients[last_level]
 
-              ## if there are less than 1/3 DLTs at that level
-              if(nDLTs[lastLevel]/nPatients[lastLevel] < 1/3)
-              {
-                  ## we could escalate, unless this is the highest
-                  ## level or the higher level was tried already
-                  ## (in which case it was not safe)
-                  if((lastLevel == length(data@doseGrid)) ||
-                     (nPatients[lastLevel+1] > 0))
-                  {
-                      nextLevel <- lastLevel
-                  } else {
-                      nextLevel <- lastLevel + 1
-                  }
-              } else if(nDLTs[lastLevel]/nPatients[lastLevel] > 1/3) {
-                  ## rate here is too high, therefore deescalate
-                  nextLevel <- lastLevel - 1
-              } else {
-                  ## otherwise: rate is 1/3 == 2/6,
-                  ## then it depends on the number of patients:
-                  ## if more than 3, then deescalate, otherwise stay.
-                  nextLevel <-
-                      if(nPatients[lastLevel] > 3)
-                          lastLevel - 1
-                      else
-                          lastLevel
-              }
-
-              ## do we stop here? only if we have no MTD
-              ## or the next level has been tried enough (more than
-              ## three patients already)
-              stopHere <-
-                  if(nextLevel == 0)
-                  {
-                      TRUE
-                  } else {
-                      nPatients[nextLevel] > 3
-                  }
-
-              ## return value and plot
-              return(list(value=
-                          if(nextLevel == 0) NA else data@doseGrid[nextLevel],
-                          stopHere=stopHere))
-          })
+    next_level <- last_level + if (DLT_rate_last_level < 1 / 3) {
```

Comment:
I would first assign the result of the if/else to a variable, e.g. `level_change` and then add that to `last_level`. otherwise too hard to read.